File: app/services/knowledge.py
"""KnowledgeService: ingestão e gestão da base RAG por agente (pontos 12, 7, 8).

Aplica a feature flag `rag_enabled` (ponto 8), valida a entrada e delega a
ingestão pesada ao `app.tasks` — que enfileira (Celery) ou roda síncrono.

Os métodos de ingestão são SÍNCRONOS e bloqueantes no caminho de fallback; o
router os executa em worker thread (`asyncio.to_thread`).
"""

from app import rag, tasks
from app.domain import AgentConfig
from app.errors import ValidationError


class KnowledgeService:
    def ingest_pdf(self, agent: AgentConfig, pdf_bytes: bytes, source_name: str) -> dict:
        self._require_rag(agent)
        if not pdf_bytes:
            raise ValidationError("Arquivo PDF vazio.")
        return tasks.submit_pdf_ingest(agent.id, pdf_bytes, source_name)

    def ingest_text(self, agent: AgentConfig, text: str, source_name: str) -> dict:
        self._require_rag(agent)
        return tasks.submit_text_ingest(agent.id, text, source_name)

    def list_sources(self, agent: AgentConfig) -> list[dict]:
        return rag.list_sources(agent.id)

    def delete_source(self, agent: AgentConfig, source_name: str) -> dict:
        return rag.delete_source(agent.id, source_name)

    @staticmethod
    def _require_rag(agent: AgentConfig) -> None:
        if not agent.rag_enabled:
            raise ValidationError(
                "Base de conhecimento desabilitada para este agente "
                "(feature flag rag_enabled=false)."
            )

File: src/server/services/knowledge.ts
/**
 * KnowledgeService (porta `app/services/knowledge.py`): ingestão e gestão da base
 * RAG por agente. Aplica a feature flag `rag_enabled` e valida a entrada.
 *
 * Ingestão SÍNCRONA (decisão da migração: Celery+Redis removidos). O I/O é async
 * em Node, então a rota apenas faz `await` — sem worker thread.
 */

import type { AgentConfig } from "../domain";
import { ValidationError } from "../errors";
import * as rag from "../rag";
import type { IngestResult, SourceInfo } from "../rag";

export class KnowledgeService {
  ingestPdf(agent: AgentConfig, pdfBytes: Uint8Array, sourceName: string): Promise<IngestResult> {
    KnowledgeService.requireRag(agent);
    if (pdfBytes.length === 0) throw new ValidationError("Arquivo PDF vazio.");
    return rag.ingestPdf(agent.id, pdfBytes, sourceName);
  }

  ingestText(agent: AgentConfig, text: string, sourceName: string): Promise<IngestResult> {
    KnowledgeService.requireRag(agent);
    return rag.ingestText(agent.id, text, sourceName);
  }

  listSources(agent: AgentConfig): Promise<SourceInfo[]> {
    return rag.listSources(agent.id);
  }

  deleteSource(agent: AgentConfig, sourceName: string): Promise<{ deleted_chunks: number; source_name: string }> {
    return rag.deleteSource(agent.id, sourceName);
  }

  private static requireRag(agent: AgentConfig): void {
    if (!agent.ragEnabled) {
      throw new ValidationError(
        "Base de conhecimento desabilitada para este agente (feature flag rag_enabled=false).",
      );
    }
  }
}
